### PR TITLE
ChibiOS: implement stm32 watchdog functionality

### PR DESCRIPTION
**Pull request by tridge**
This adds a one second watchdog to ChibiOS builds, enabled with BRD_OPTIONS=1
This guarantees that if the main loop doesn't run for 1 second then the board will reset. When it resets it skips gyro cal, baro cal and airspeed cal, allowing for very fast boot

### CODE REVIEW

**Comment by WickedShell** [comment]
In the event of a watchdog reboot we probably want to grab the home location back out of mission storage, and lock that in as well, that way the distance from home/rtl functionality will work. (This may need different logic on other vehicles then plane admittedly). Ideally we'd actually stuff the EKF origin as well into the FRAM, but we don't have a slot waiting for it....

**Comment by proficnc** [comment]
We can also use battery backed ram.

**Comment by WickedShell** [comment]
> We can also use battery backed ram.

I hesitate to, on the basis that if we need a watchdog everything in RAM should be considered as highly questionable. We could easily have been walking the stack and resetting everything to fun random values, FRAM is less likely to be corrupted by an out of control loop (particularly if checksummed).

**Comment by davidbuzz** [comment]
Does the operator also get notified by a ground station message after reboot?

**Comment by tridge** [comment]
need to cope with log delete timeout

**Comment by tridge** [comment]
should add status bit for watchdog reset

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp [hunk 1]
```diff
@@ -472,4 +480,24 @@ bool Scheduler::thread_create(AP_HAL::MemberProc proc, const char *name, uint32_
...
+  A value of zero cancels the previous expected delay
+*/
+void Scheduler::expect_delay_ms(uint32_t ms)
+{
+    if (!in_main_thread()) {
+        // only for main thread
+        return;
+    }
+    if (ms == 0) {
+        expect_delay_start = 0;
```
Comment (WickedShell) [inline libraries/AP_HAL_ChibiOS/Scheduler.cpp]: `expect_delay_length` should be reset here, or you get a extra round of pats on time wrap.

**Comment by tridge** [comment]
add status bits to internal_error

**Comment by tridge** [comment]
should remember safety state?

File: libraries/AP_HAL_ChibiOS/hwdef/common/board.c [hunk 2]
```diff
@@ -233,6 +234,10 @@ void __early_init(void) {
 void __late_init(void) {
   halInit();
   chSysInit();
+  stm32_watchdog_save_reason();
+#ifndef HAL_BOOTLOADER_BUILD
+  stm32_watchdog_clear_reason();
```
Comment (WickedShell) [inline libraries/AP_HAL_ChibiOS/hwdef/common/board.c]: Wouldn't we want to do this normally? IE if we had an old bootloader that wasn't updated?

**Comment by tridge** [comment]
use BRD_WATCHDOG parameter for time to reset?

**Comment by ChristopherOlson** [comment]
@tridge as we discussed for helicopter, if the board reboots with cyclic and torque (rudder) control, regardless of safety switch settings, and would default to Acro flight mode regardless of the flight mode switch position, this should be pretty much a non-issue for heli. It would be like any other in-fight engine failure. As long as the pilot has autorotation control the helicopter can be safely landed. Or at least reduce the energy on impact to minimize damage to the machine. Both of which are much better than having it fly off in uncontrolled state.

I'll ping @bnsgeyer in case he has extra comments. I voted for defaulting to Acro due to the likelihood of the attitude solution not being accurate on fast reboot.

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Scheduler.cpp]: i don't think that's right. It checks for zero.

File: libraries/AP_HAL_ChibiOS/hwdef/common/board.c (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/common/board.c]: bootloader must not clear the reason, or main fw can't see it

**Comment by Jaaaky** [comment]
You should also add IOMCU watchdog. I'd many on-ground - and 1 during flight - recent cases where IOMCU got frozen due to some bad ICE rpm sensor.
FMU was working, sending status and updating info, but all IO ports are sending 1500 PWM.
FMU could detect that all ports are having 1500 PWM, as it was displayed on GCS RCouts.
So in such case IOMCU should reset; however I can remember that GCS reboot command was not working.

**Comment by peterbarker** [comment]
On Sun, 21 Apr 2019, Jaaaky wrote:

> You should also add IOMCU watchdog. I'd many on-ground - and 1 during flight - recent cases where IOMCU got frozen due to some bad ICE rpm sensor.

How is the RPM sensor faulty?
What is the RPM sensor connected to?
Which flightboard?
Which version of ArduPilot?

> FMU was working, sending status and updating info, but all IO ports are sending 1500 PWM.

Any outputs on the aux ports still known working when in this state?

> FMU could detect that all ports are having 1500 PWM, as it was displayed on GCS RCouts.

Depends on how you're seeing that "1500"; IIRC we read back the output 
values from the IOMCU for display....

> So in such case IOMCU should reset; however I can remember that GCS reboot command was not working.

Were you armed at the time?  We now refuse to reboot when armed (at least 
on master).

It is planned for this going to to into the IOMCU code at some stage.

**Comment by tridge** [comment]
@Jaaaky I've added detection of watchdog reset in IOMCU and tested it. It also triggers a resend of the mixer data from FMU, and if IOMCU safety was off when watchdog fires then it also disables safety reboot

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 3]
```diff
@@ -157,7 +190,16 @@ AP_AHRS_DCM::reset(bool recover_eulers)
     // if the caller wants us to try to recover to the current
     // attitude then calculate the dcm matrix from the current
     // roll/pitch/yaw values
-    if (recover_eulers && !isnan(roll) && !isnan(pitch) && !isnan(yaw)) {
+    if (hal.util->was_watchdog_reset() &&
+        hal.util->get_backup_attitude(roll_sensor, pitch_sensor, yaw_sensor) &&
+        AP_HAL::millis() < 10000) {
```
Comment (WickedShell) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp]: If CAN timeouts ever grow close to this, this won't ever trigger.

File: libraries/AP_IOMCU/AP_IOMCU.cpp [hunk 4]
```diff
@@ -884,4 +891,21 @@ const char *AP_IOMCU::get_rc_protocol(void)
...
+void AP_IOMCU::handle_repeated_failures(void)
+{
+    detected_io_reset = true;
+    AP::internalerror().error(AP_InternalError::error_t::iomcu_reset);
+    // we need to ensure the mixer data and the rates are sent over to
+    // the IOMCU
+    if (mixing.enabled) {
+        trigger_event(IOEVENT_MIXING);
+    }
+    trigger_event(IOEVENT_SET_RATES);
```
Comment (WickedShell) [inline libraries/AP_IOMCU/AP_IOMCU.cpp]: Default rate needs to be set again.

File: libraries/AP_IOMCU/AP_IOMCU.cpp [hunk 5]
```diff
@@ -420,6 +421,9 @@ bool AP_IOMCU::read_registers(uint8_t page, uint8_t offset, uint8_t count, uint1
         return false;
     }
     memcpy(regs, pkt.regs, count*2);
+    if (protocol_fail_count > 3) {
```
Comment (WickedShell) [inline libraries/AP_IOMCU/AP_IOMCU.cpp]: Seems a bit tight given that we talked about shifting to a counter of 5 to be considered unhealthy.

File: libraries/AP_IOMCU/AP_IOMCU.cpp [hunk 6]
```diff
@@ -485,6 +489,9 @@ bool AP_IOMCU::write_registers(uint8_t page, uint8_t offset, uint8_t count, cons
         protocol_fail_count++;
         return false;
     }
+    if (protocol_fail_count > 3) {
```
Comment (WickedShell) [inline libraries/AP_IOMCU/AP_IOMCU.cpp]: The 3 should really be a define.

**Comment by Jaaaky** [comment]
> On Sun, 21 Apr 2019, Jaaaky wrote: You should also add IOMCU watchdog. I'd many on-ground - and 1 during flight - recent cases where IOMCU got frozen due to some bad ICE rpm sensor.
> How is the RPM sensor faulty? What is the RPM sensor connected to? Which flightboard? Which version of ArduPilot?

I could find a telemetry log of one case. So I can give answers according:
Actually it was an cheap ICE with high EMI noise - discovered this later -, so the RPM sensors acted like EMI amplifier that caused repetitive board freeze.
It was connected to AUX but was not configured yet; i.e. params were not enabled.
It was Pixhawk1 board v2.4.8
ArduPlane master commit b7c88da67c 

> FMU was working, sending status and updating info, but all IO ports are sending 1500 PWM.
> Any outputs on the aux ports still known working when in this state?

I didn't have any AUX ports usage then. But IMU/GPS/baro/board voltage was indeed working and updating attitude and status correctly. I could change flight mode through telemetry. Plane was ARMED and SAFETY armed after passing all checks successfully.
All configured servo outputs were set to 1500 except servo3 - throttle - which was at servo3 trim/min PWM.
Also servo rail voltage stopped updating, while "board voltage" - through power module was updating.

> FMU could detect that all ports are having 1500 PWM, as it was displayed on GCS RCouts.
> Depends on how you're seeing that "1500"; IIRC we read back the output values from the IOMCU for display....

Yeah, but physically all control surfaces were suck at trim - 1500 PWM - and do not move even in FBWA/AUTO modes while moving the plane. Auto mode commanded 100% throttle, I could see it in ch3percet in MP, but throttle didn't move at all.
 
> So in such case IOMCU should reset; however I can remember that GCS reboot command was not working.
> Were you armed at the time? We now refuse to reboot when armed (at least on master). It is planned for this going to to into the IOMCU code at some stage.

Yeah I'm aware of the no-reboot-if-armed change, but it was not applied yet.

**Comment by Jaaaky** [comment]
@tridge Should the watchdog commits be useful in the mentioned case or some further change should be done?
I can try to reproduce the problem and check if you think it is ready for such handling case.

**Pull request merged**